enhancements/network/user-defined-network-segmentation.md

```diff
@@ -357,11 +359,106 @@ block/allow cluster egress or ingress traffic will still be enforced.
+- Cluster scoped CRD - represent user request for creating cluster scoped OVN network, enables cross-namespace networking.
+  - Shall be defined with `cluster` scope
+
+The CRD instance There should be will be a finalizer on the CRDs, so that upon deletion OVN-Kubernetes can validate that there are no pods still using this network.
+If there are pods still attached to this network, the network will not be removed.
+
+##### spec
+
+The CRDs spec defines as follows:
+
+| Field name     | Description                                                                                                                                                                                                                                                                                                                                                                                        | optional |
+|----------------|----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|----------|
+| Topology       | The topological configuration for the network. Must be one of `layer2` or `layer3`.                                                                                                                                                                                                                                                                                                                | No       |
+| Role           | Select the network role in the pod, either `primary` or `secondary`.                                                                                                                                                                                                                                                                                                                               | No       |
+| MTU            | The maximum transmission unit (MTU).<br/>The default value, 1300, is automatically set by the kernel                                                                                                                                                                                                                                                                                               | Yes      |
+| Subnets        | The subnet to use for the network across the cluster.<br/>E.g. 10.100.200.0/24.<br/>IPv6 (2001:DBB::/64) and dual-stack (192.168.100.0/24,2001:DBB::/64) subnets are supported.<br/>When omitted, the logical switch implementing the network only provides layer 2 communication, and users must configure IP addresses.<br/>Port security only prevents MAC spoofing if the subnets are omitted. | Yes      |
```

Comment:
I think the CR should expose the least of options, although it sounds mandatory, do you think we can avoid exposing the join subnet?

Could there be a way for the backing CRDs controller to realize what the join subnet should be?

Reply:
OH! having join subnet via CR would be better, it definitely belongs here specially if podsubnet is via here.

Perhaps then for NAD config approach we can do the same and add joinSubnet as a config field to the NAD? that will save us from having to do a new API change in ocp/api for join subnets...

great idea tim..

Reply:
Summary of offline discussion about join subnet, for primary networks the join subnet can be set as optional field, in case no join subnet is set, a default one should be set by the controller.
Advance users can choose the CIRD to use for the join subnet by specifying the field.

Reply:
A few things we need to make clear about join subnet in the enhancement:
1. Join subnet is not applicable to secondary networks. Join subnet is used for north/south traffic, and secondary networks will not support north/south at this time. We don't need to stop a user from configuring the join subnet for a seondary network CR, but it wont do anything.
2. The join subnet specified in the CR cannot overlap with the cluster default network join subnet. This subnet is configured in OVN-Kubernetes via: https://github.com/ovn-org/ovn-kubernetes/blob/master/go-controller/pkg/config/config.go#L441
with defaults here: https://github.com/ovn-org/ovn-kubernetes/blob/master/go-controller/pkg/config/config.go#L138
You can simply add to this enhancement:
"The join subnet shall not overlap with the configured cluster default network join subnet. If there is overlap, OVN-Kubernetes will report an error status to the network CR"
3. Join subnet specified in the CR may overlap with other UDNs. There is no conflict there. According to Surya's PR, the default subnet chosen for a UDN will be a constant declared at:
https://github.com/ovn-org/ovn-kubernetes/pull/4507/commits/05e13058d038886aebdf1d28e4f841a7fc0b4889#diff-ee7e8cb2c18f89a20eaaf4c10d14e404559cc7c7dbaabe1801566ba57e2b5ed7R113

Reply:
Thanks for clarifying, added text about join subnet validation.